File: sdks/python/conformance_runner.py
#!/usr/bin/env python3
"""Protocol conformance runner for the beachcomber Python SDK.

Loads fixtures from tests/conformance/ relative to the repo root, spawns
a fresh daemon for each fixture group, drives the ops through the Python
client, and asserts the expected outcomes.

Usage::

    COMB_BIN=/path/to/comb python sdks/python/conformance_runner.py

The runner exits with code 0 if all fixtures pass and code 1 if any fail
or if COMB_BIN is unset / the binary cannot be found.
"""

from __future__ import annotations

import json
import os
import pathlib
import shutil
import signal
import socket
import subprocess
import sys
import tempfile
import time
from typing import Any, Optional

# ---------------------------------------------------------------------------
# Path setup — make the SDK importable when run as a script.
# ---------------------------------------------------------------------------

_SDK_DIR = pathlib.Path(__file__).parent
sys.path.insert(0, str(_SDK_DIR))

from libbeachcomber.client import Client  # noqa: E402
from libbeachcomber.exceptions import CombError, ServerError  # noqa: E402
from libbeachcomber.types import IntrospectSubject  # noqa: E402

# ---------------------------------------------------------------------------
# Fixture discovery
# ---------------------------------------------------------------------------

_REPO_ROOT = _SDK_DIR.parent.parent
_CONFORMANCE_DIR = _REPO_ROOT / "tests" / "conformance"


def discover_fixtures() -> list[pathlib.Path]:
    """Return all *.json fixture files sorted by path."""
    if not _CONFORMANCE_DIR.exists():
        print(f"ERROR: conformance directory not found: {_CONFORMANCE_DIR}", file=sys.stderr)
        sys.exit(1)
    fixtures = sorted(_CONFORMANCE_DIR.rglob("*.json"))
    return fixtures


# ---------------------------------------------------------------------------
# Daemon lifecycle
# ---------------------------------------------------------------------------


def _wait_for_socket(sock_path: str, timeout: float = 5.0) -> bool:
    """Poll until the Unix socket exists and accepts connections."""
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if os.path.exists(sock_path):
            try:
                s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                s.settimeout(0.1)
                s.connect(sock_path)
                s.close()
                return True
            except OSError:
                pass
        time.sleep(0.05)
    return False


class DaemonProcess:
    """Manages a comb daemon subprocess for one conformance run."""

    def __init__(self, comb_bin: str) -> None:
        self._comb_bin = comb_bin
        self._tmpdir = tempfile.mkdtemp(prefix="bcconf_")
        self.socket_path = os.path.join(self._tmpdir, "comb.sock")
        self._config_path = os.path.join(self._tmpdir, "config.toml")
        self._proc: Optional[subprocess.Popen[bytes]] = None
        self._log_path = os.path.join(self._tmpdir, "daemon.log")

        # Write a minimal config pointing at our temp socket.
        with open(self._config_path, "w") as f:
            f.write(f'[daemon]\nsocket_path = "{self.socket_path}"\n')

    def start(self) -> None:
        """Start the daemon process."""
        env = os.environ.copy()
        env["XDG_CONFIG_HOME"] = self._tmpdir  # isolate from user config

        with open(self._log_path, "wb") as log:
            self._proc = subprocess.Popen(
                [self._comb_bin, "daemon", "--socket", self.socket_path],
                stdout=log,
                stderr=log,
                env=env,
            )

        if not _wait_for_socket(self.socket_path, timeout=8.0):
            self.stop()
            log_tail = ""
            try:
                with open(self._log_path) as f:
                    log_tail = f.read()[-2000:]
            except OSError:
                pass
            raise RuntimeError(
                f"daemon did not start within 8s (socket: {self.socket_path})\n"
                f"Log tail:\n{log_tail}"
            )

    def stop(self) -> None:
        """Terminate the daemon process and clean up."""
        if self._proc is not None:
            try:
                self._proc.send_signal(signal.SIGTERM)
                self._proc.wait(timeout=3.0)
            except (ProcessLookupError, subprocess.TimeoutExpired):
                try:
                    self._proc.kill()
                except ProcessLookupError:
                    pass
            self._proc = None
        shutil.rmtree(self._tmpdir, ignore_errors=True)

    def make_client(self) -> Client:
        """Return a Client connected to this daemon's socket."""
        return Client(socket_path=self.socket_path, timeout=5.0)


# ---------------------------------------------------------------------------
# Op dispatch helpers
# ---------------------------------------------------------------------------


def _run_op(
    client: Client,
    op: str,
    args: dict[str, Any],
    resolve_ctx: Optional[dict[str, Any]] = None,
) -> Any:
    """Run a single op against the client.

    Returns the raw response data (or None for ops that return nothing),
    or raises CombError / ServerError on failure.

    `resolve_ctx` (only consulted for the `resolve` op) carries the
    fixture's `virtual`/`env`/`cwd` blocks — see
    tests/conformance/README.md.
    """
    if op == "resolve":
        assert resolve_ctx is not None
        key = args.get("key", "")
        return client.resolve(
            key,
            cwd=resolve_ctx["cwd"],
            env=resolve_ctx["env"] or None,
            overrides=resolve_ctx["virtual"] or None,
        )
    if op == "get":
        return client.get(**args)
    if op == "refresh":
        client.refresh(**args)
        return None
    if op == "put":
        client.put(**args)
        return None
    if op == "status":
        return client.status()
    if op == "hello":
        return client.hello()
    if op == "context":
        # Context is session-level; run through a session to avoid issues.
        with client.session() as session:
            session.set_context(args.get("path", ""))
        return None
    if op == "watch":
        key = args.get("key", "")
        path = args.get("path")
        with client.watch(key, path=path) as stream:
            event = stream.next_event()
        return event
    if op == "introspect":
        subject_str = args.get("subject", "daemon")
        try:
            subject = IntrospectSubject(subject_str)
        except ValueError:
            raise ValueError(f"unknown introspect subject: {subject_str!r}")
        duration_secs = args.get("duration_secs")
        return client.introspect(subject, duration_secs=duration_secs)
    raise ValueError(f"unknown op: {op!r}")


# Ops this runner's binding can execute. A fixture using any op outside this
# set must be skipped, not failed — the binding doesn't implement it yet.
_SUPPORTED_OPS = {
    "hello",
    "get",
    "refresh",
    "put",
    "status",
    "context",
    "watch",
    "introspect",
    "resolve",
}


def _unsupported_op(fixture: dict[str, Any]) -> Optional[str]:
    """Return the first op in the fixture that this runner can't execute, or None."""
    ops = [s.get("op") for s in fixture.get("setup", [])]
    ops.append(fixture["test"]["op"])
    for op in ops:
        if op not in _SUPPORTED_OPS:
            return op
    return None


# ---------------------------------------------------------------------------
# Assertion helpers
# ---------------------------------------------------------------------------


def _check_expect(
    expect: dict[str, Any],
    result: Any,
    error: Optional[str],
    fixture_name: str,
) -> list[str]:
    """Check the expectation dict against the result/error.

    Returns a list of failure messages (empty on success).
    """
    failures: list[str] = []
    status = expect.get("status", "ok")

    if status == "ok" or status == "hit":
        if error is not None:
            failures.append(f"expected ok/hit but got error: {error!r}")
            return failures

        # For get results (CombResult).
        from libbeachcomber.result import CombResult
        from libbeachcomber.types import HelloInfo, IntrospectResponse, WatchEvent

        if status == "hit":
            if isinstance(result, CombResult):
                if not result.is_hit:
                    failures.append("expected cache hit but got miss (data is None)")
            elif isinstance(result, WatchEvent):
                if result.data is None:
                    failures.append("expected watch hit but data is None")
            else:
                failures.append(f"expected CombResult or WatchEvent for hit, got {type(result).__name__}")

        # data_type check.
        data_type = expect.get("data_type")
        if data_type is not None:
            actual_data = _extract_data(result)
            if not _check_data_type(actual_data, data_type):
                failures.append(
                    f"expected data_type={data_type!r} but data was {actual_data!r} "
                    f"(type {type(actual_data).__name__})"
                )

        # data_equals check.
        if "data_equals" in expect:
            expected_val = expect["data_equals"]
            actual_data = _extract_data(result)
            if actual_data != expected_val:
                failures.append(
                    f"expected data == {expected_val!r} but got {actual_data!r}"
                )

        # data_as_text check.
        if "data_as_text" in expect:
            expected_text = str(expect["data_as_text"])
            actual_data = _extract_data(result)
            if str(actual_data) != expected_text:
                failures.append(
                    f"expected data as text {expected_text!r} but got {str(actual_data)!r}"
                )

        # data_contains_field check.
        if "data_contains_field" in expect:
            field = expect["data_contains_field"]
            actual_data = _extract_data(result)
            if not isinstance(actual_data, dict) or field not in actual_data:
                failures.append(
                    f"expected data to contain field {field!r} but data was {actual_data!r}"
                )

        # age_ms_present check.
        if expect.get("age_ms_present"):
            age = _extract_age_ms(result)
            if age is None:
                failures.append("expected age_ms to be present but it was None/missing")

    elif status == "error":
        if error is None:
            failures.append(f"expected an error but op succeeded with result: {result!r}")
            return failures
        error_contains = expect.get("error_contains")
        if error_contains and error_contains not in error:
            failures.append(
                f"expected error to contain {error_contains!r} but got: {error!r}"
            )

    elif status == "miss":
        if error is not None:
            failures.append(f"expected miss but got error: {error!r}")
        else:
            from libbeachcomber.result import CombResult
            if isinstance(result, CombResult) and result.is_hit:
                failures.append(f"expected cache miss but got hit with data: {result.data!r}")

    else:
        failures.append(f"unknown expect status: {status!r}")

    return failures


def _extract_data(result: Any) -> Any:
    """Pull the data value out of typed result objects."""
    from libbeachcomber.result import CombResult
    from libbeachcomber.types import HelloInfo, IntrospectResponse, WatchEvent

    if isinstance(result, CombResult):
        return result.data
    if isinstance(result, WatchEvent):
        return result.data
    if isinstance(result, HelloInfo):
        return {"protocol_version": result.protocol_version, "daemon_version": result.daemon_version}
    if isinstance(result, IntrospectResponse):
        if result.daemon is not None:
            import dataclasses
            return dataclasses.asdict(result.daemon)
        return result.other
    return result


def _extract_age_ms(result: Any) -> Optional[int]:
    from libbeachcomber.result import CombResult
    from libbeachcomber.types import WatchEvent

    if isinstance(result, CombResult):
        return result.age_ms
    if isinstance(result, WatchEvent):
        return result.age_ms
    return None


def _check_data_type(data: Any, expected_type: str) -> bool:
    if expected_type == "object":
        return isinstance(data, dict)
    if expected_type == "array":
        return isinstance(data, list)
    if expected_type == "string":
        return isinstance(data, str)
    if expected_type == "number":
        return isinstance(data, (int, float)) and not isinstance(data, bool)
    if expected_type == "boolean":
        return isinstance(data, bool)
    if expected_type == "null":
        return data is None
    return False


# ---------------------------------------------------------------------------
# Run a single fixture
# ---------------------------------------------------------------------------


def run_fixture(fixture_path: pathlib.Path, client: Client) -> tuple[str, str]:
    """Run a single fixture against the given client.

    Returns (status, message) where status is "pass", "fail", or "skip".
    """
    with open(fixture_path) as f:
        fixture = json.load(f)

    name = fixture.get("name", fixture_path.stem)

    unsupported = _unsupported_op(fixture)
    if unsupported is not None:
        return "skip", f"unsupported op {unsupported!r}"

    with tempfile.TemporaryDirectory(prefix="bcconf_cwd_") as default_cwd:
        resolve_ctx = {
            "virtual": fixture.get("virtual", {}),
            "env": fixture.get("env", {}),
            "cwd": fixture.get("cwd", default_cwd),
        }

        # Run setup ops (ignore their results; failures are fatal for the fixture).
        for setup_op in fixture.get("setup", []):
            op = setup_op["op"]
            args = setup_op.get("args", {})
            try:
                _run_op(client, op, args, resolve_ctx)
            except Exception as exc:
                return "fail", f"setup op {op!r} failed: {exc}"

        # Run the test op.
        test = fixture["test"]
        op = test["op"]
        args = test.get("args", {})
        expect = fixture.get("expect", {})

        result = None
        error: Optional[str] = None

        try:
            result = _run_op(client, op, args, resolve_ctx)
        except ServerError as exc:
            error = exc.message
        except CombError as exc:
            error = str(exc)
        except Exception as exc:
            return "fail", f"unexpected exception running {op!r}: {type(exc).__name__}: {exc}"

        failures = _check_expect(expect, result, error, name)
        if failures:
            msg = "; ".join(failures)
            return "fail", f"{name}: FAIL — {msg}"

        return "pass", f"{name}: ok"


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------


def main() -> int:
    comb_bin = os.environ.get("COMB_BIN", shutil.which("comb") or "")
    if not comb_bin:
        print(
            "ERROR: COMB_BIN environment variable is not set and 'comb' is not in PATH.\n"
            "Set COMB_BIN=/path/to/comb to run the conformance suite.",
            file=sys.stderr,
        )
        return 1

    if not os.path.isfile(comb_bin) or not os.access(comb_bin, os.X_OK):
        print(f"ERROR: COMB_BIN={comb_bin!r} is not an executable file.", file=sys.stderr)
        return 1

    fixtures = discover_fixtures()
    if not fixtures:
        print("WARNING: No conformance fixtures found.", file=sys.stderr)
        return 0

    print(f"Found {len(fixtures)} fixture(s) in {_CONFORMANCE_DIR}")
    print(f"Using daemon: {comb_bin}\n")

    passed = 0
    failed = 0
    skipped = 0
    errors: list[str] = []

    # A fresh daemon per fixture, per tests/conformance/README.md's
    # isolation guarantee ("Each fixture runs against a fresh daemon
    # instance") — a shared daemon would leak cache state (e.g. `put`s from
    # one fixture's `setup` visible to the next fixture's `resolve` cache
    # refs) across fixtures that happen to reuse a key.
    for fixture_path in fixtures:
        rel = fixture_path.relative_to(_CONFORMANCE_DIR)

        daemon = DaemonProcess(comb_bin)
        try:
            daemon.start()
        except RuntimeError as exc:
            print(f"ERROR: Failed to start daemon for {rel}: {exc}", file=sys.stderr)
            return 1

        try:
            client = daemon.make_client()
            status, msg = run_fixture(fixture_path, client)
        finally:
            daemon.stop()

        if status == "skip":
            print(f"  SKIP {rel}: {msg}")
            skipped += 1
            continue
        status_label = "PASS" if status == "pass" else "FAIL"
        print(f"  [{status_label}] {rel}: {msg.split(': ', 1)[-1]}")
        if status == "pass":
            passed += 1
        else:
            failed += 1
            errors.append(str(rel) + ": " + msg)

    print(f"\n{passed + failed + skipped} fixtures: {passed} passed, {failed} failed, {skipped} skipped")

    if errors:
        print("\nFailures:")
        for e in errors:
            print(f"  - {e}")
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdks/node/conformance_runner.js
#!/usr/bin/env node
/**
 * Protocol conformance runner for the beachcomber Node.js SDK.
 *
 * Loads fixtures from tests/conformance/ relative to the repo root, spawns
 * a fresh daemon per fixture, drives the ops through the Node SDK, and
 * asserts the expected outcomes.
 *
 * Usage:
 *   COMB_BIN=/path/to/comb node sdks/node/conformance_runner.js
 *
 * Exits 0 if all fixtures pass, 1 otherwise.
 *
 * The runner validates typed SDK results, not raw JSON. For example, an
 * introspect{daemon} assertion on data_contains_field:"pid" is checked via
 * DaemonHealth.pid, not by reparsing the raw JSON wire response.
 */

'use strict';

import fs from 'fs';
import net from 'net';
import path from 'path';
import os from 'os';
import { spawnSync, spawn } from 'child_process';

// ---------------------------------------------------------------------------
// Path setup — import the SDK from compiled dist/ or via tsx if available.
// ---------------------------------------------------------------------------

const SDK_DIR = path.dirname(new URL(import.meta.url).pathname);
const REPO_ROOT = path.resolve(SDK_DIR, '..', '..');
const CONFORMANCE_DIR = path.join(REPO_ROOT, 'tests', 'conformance');
const DIST_CLIENT = path.join(SDK_DIR, 'dist', 'client.js');
const DIST_ERRORS = path.join(SDK_DIR, 'dist', 'errors.js');

if (!fs.existsSync(DIST_CLIENT)) {
  console.error(`ERROR: dist/client.js not found. Run 'npm run build' first in sdks/node/`);
  process.exit(1);
}

// Dynamic import is async; use top-level async IIFE.
(async () => {
  const { Client, WatchStream } = await import(DIST_CLIENT);
  const { ServerError } = await import(DIST_ERRORS);

  // ---------------------------------------------------------------------------
  // Fixture discovery
  // ---------------------------------------------------------------------------

  function discoverFixtures() {
    if (!fs.existsSync(CONFORMANCE_DIR)) {
      console.error(`ERROR: conformance directory not found: ${CONFORMANCE_DIR}`);
      process.exit(1);
    }
    const fixtures = [];
    for (const opDir of fs.readdirSync(CONFORMANCE_DIR)) {
      const opPath = path.join(CONFORMANCE_DIR, opDir);
      if (!fs.statSync(opPath).isDirectory()) continue;
      for (const file of fs.readdirSync(opPath)) {
        if (!file.endsWith('.json')) continue;
        const filePath = path.join(opPath, file);
        const text = fs.readFileSync(filePath, 'utf8');
        const v = JSON.parse(text);
        fixtures.push({
          name: v.name,
          description: v.description,
          setup: Array.isArray(v.setup) ? v.setup : [],
          test: v.test,
          expect: v.expect,
          sourcePath: filePath,
        });
      }
    }
    return fixtures;
  }

  // ---------------------------------------------------------------------------
  // Daemon lifecycle
  // ---------------------------------------------------------------------------

  const COMB_BIN = process.env['COMB_BIN'] || 'comb';

  function waitForSocket(sockPath, timeoutMs = 5000) {
    const deadline = Date.now() + timeoutMs;
    return new Promise((resolve) => {
      function attempt() {
        if (Date.now() > deadline) {
          resolve(false);
          return;
        }
        const s = new net.Socket();
        s.setTimeout(100);
        s.connect(sockPath, () => {
          s.destroy();
          resolve(true);
        });
        s.on('error', () => {
          setTimeout(attempt, 50);
        });
        s.on('timeout', () => {
          s.destroy();
          setTimeout(attempt, 50);
        });
      }
      attempt();
    });
  }

  async function spawnDaemon() {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'beachcomber-conform-'));
    const sockPath = path.join(tmpDir, 'comb.sock');

    const proc = spawn(COMB_BIN, ['daemon', '--socket', sockPath], {
      stdio: ['ignore', 'pipe', 'pipe'],
      env: { ...process.env, COMB_SOCKET: sockPath },
    });

    proc.stdout.on('data', () => {});
    proc.stderr.on('data', () => {});

    const ready = await waitForSocket(sockPath, 5000);
    if (!ready) {
      proc.kill('SIGKILL');
      fs.rmSync(tmpDir, { recursive: true, force: true });
      throw new Error(`daemon did not start within 5s (COMB_BIN=${COMB_BIN})`);
    }

    return {
      proc,
      sockPath,
      tmpDir,
      stop() {
        try { proc.kill('SIGTERM'); } catch {}
        try { fs.rmSync(tmpDir, { recursive: true, force: true }); } catch {}
      },
    };
  }

  // ---------------------------------------------------------------------------
  // Op dispatch
  // ---------------------------------------------------------------------------

  /**
   * Canonical response shape used by expectation checking.
   * Mirrors the Rust conformance runner's CanonicalResponse.
   */
  function makeOk({ data = null, dataAsText = null, ageMs = null, stale = null } = {}) {
    return { ok: true, data, dataAsText, ageMs, stale, error: null };
  }

  function makeError(message) {
    return { ok: false, data: null, dataAsText: null, ageMs: null, stale: null, error: message };
  }

  function valueAsText(v) {
    if (v === null || v === undefined) return null;
    if (typeof v === 'string') return v;
    if (typeof v === 'number' || typeof v === 'boolean') return String(v);
    // Objects: not directly as text.
    return null;
  }

  async function runOp(client, descriptor) {
    const { op, args } = descriptor;

    try {
      switch (op) {
        case 'hello': {
          const info = await client.hello();
          const data = {
            protocol_version: info.protocolVersion,
            daemon_version: info.daemonVersion,
          };
          return makeOk({ data });
        }

        case 'get': {
          const key = args.key || '';
          const pathArg = args.path;
          const result = await client.get(key, pathArg);
          if (result.isHit) {
            return makeOk({
              data: result.data,
              dataAsText: result.getString(),
              ageMs: result.ageMs,
              stale: result.stale,
            });
          } else {
            return makeOk();
          }
        }

        case 'refresh': {
          const key = args.key || '';
          const pathArg = args.path;
          await client.refresh(key, pathArg);
          return makeOk();
        }

        case 'context': {
          // Context is session-based; use a fresh session.
          const session = await client.session();
          try {
            await session.setContext(args.path || '/tmp');
            return makeOk();
          } finally {
            session.close();
          }
        }

        case 'put': {
          const key = args.key || '';
          const data = args.data;
          const ttl = args.ttl;
          const pathArg = args.path;
          await client.put(key, data, { ttl, path: pathArg });
          return makeOk();
        }

        case 'status': {
          const rows = await client.status();
          const arr = rows.map((r) => ({
            provider: r.provider,
            field: r.field,
            path: r.path,
            value: r.value,
            age_ms: r.ageMs,
            stale: r.stale,
          }));
          return makeOk({ data: arr });
        }

        case 'watch': {
          const key = args.key || '';
          const pathArg = args.path;
          const stream = await client.watch(key, pathArg);
          let event = null;
          try {
            for await (const ev of stream) {
              event = ev;
              stream.close();
              break;
            }
          } catch (err) {
            if (!event) {
              return makeError(err instanceof Error ? err.message : String(err));
            }
            // Had an event before close threw — treat as success.
          }
          if (event) {
            return makeOk({
              data: event.data,
              dataAsText: valueAsText(event.data),
              ageMs: event.ageMs,
              stale: event.stale,
            });
          }
          return makeOk();
        }

        case 'introspect': {
          const subject = args.subject || 'daemon';
          const durationSecs = args.duration_secs;
          const resp = await client.introspect(subject, { durationSecs });
          if (resp.subject === 'daemon') {
            const h = resp.daemon;
            const data = {
              pid: h.pid,
              version: h.version,
              uptime_secs: h.uptimeSecs,
              socket_path: h.socketPath,
              config_path: h.configPath,
              requests_total: h.requestsTotal,
              in_flight: h.inFlight,
              active_watchers: h.activeWatchers,
              cache_entries: h.cacheEntries,
            };
            return makeOk({ data });
          } else {
            return makeOk({ data: resp.other });
          }
        }

        default:
          throw new Error(`unknown op in fixture: ${op}`);
      }
    } catch (err) {
      return makeError(err instanceof Error ? err.message : String(err));
    }
  }

  // ---------------------------------------------------------------------------
  // Expectation checking
  // ---------------------------------------------------------------------------

  function jsonTypeOf(v) {
    if (v === null) return 'null';
    if (Array.isArray(v)) return 'array';
    if (typeof v === 'object') return 'object';
    if (typeof v === 'string') return 'string';
    if (typeof v === 'number') return 'number';
    if (typeof v === 'boolean') return 'bool';
    return 'unknown';
  }

  function deepEqual(a, b) {
    return JSON.stringify(a) === JSON.stringify(b);
  }

  function checkExpect(fixture, resp) {
    const expect = fixture.expect;
    const failures = [];

    // status
    if (expect.status) {
      switch (expect.status) {
        case 'ok':
          if (!resp.ok) failures.push(`status=ok expected but got error: ${resp.error}`);
          break;
        case 'hit':
          if (!resp.ok) failures.push(`status=hit expected but got error: ${resp.error}`);
          else if (resp.data === null) failures.push('status=hit expected but data was absent');
          break;
        case 'miss':
          if (!resp.ok) failures.push(`status=miss expected but got error: ${resp.error}`);
          else if (resp.data !== null) failures.push('status=miss expected but data was present');
          break;
        case 'error':
          if (resp.ok) failures.push('status=error expected but response was ok');
          break;
        default:
          failures.push(`unknown status: ${expect.status}`);
      }
    }

    // data_type
    if (expect.data_type !== undefined) {
      const actual = jsonTypeOf(resp.data);
      if (actual !== expect.data_type) {
        failures.push(`data_type=${expect.data_type} expected but got ${actual}: data=${JSON.stringify(resp.data)}`);
      }
    }

    // data_equals
    if (expect.data_equals !== undefined) {
      if (!deepEqual(resp.data, expect.data_equals)) {
        failures.push(`data_equals failed: expected ${JSON.stringify(expect.data_equals)}, got ${JSON.stringify(resp.data)}`);
      }
    }

    // data_as_text
    if (expect.data_as_text !== undefined) {
      const actual = resp.dataAsText ?? '';
      if (actual !== expect.data_as_text) {
        failures.push(`data_as_text=${JSON.stringify(expect.data_as_text)} expected but got ${JSON.stringify(actual)}`);
      }
    }

    // data_contains_field
    if (expect.data_contains_field !== undefined) {
      if (
        resp.data === null ||
        typeof resp.data !== 'object' ||
        Array.isArray(resp.data) ||
        !(expect.data_contains_field in resp.data)
      ) {
        failures.push(`data_contains_field=${expect.data_contains_field} failed: data=${JSON.stringify(resp.data)}`);
      }
    }

    // data_field_equals
    if (expect.data_field_equals !== undefined) {
      const { field, value } = expect.data_field_equals;
      if (
        resp.data === null ||
        typeof resp.data !== 'object' ||
        Array.isArray(resp.data)
      ) {
        failures.push(`data_field_equals: data is not an object: ${JSON.stringify(resp.data)}`);
      } else if (!deepEqual(resp.data[field], value)) {
        failures.push(`data_field_equals failed for ${field}: expected ${JSON.stringify(value)}, got ${JSON.stringify(resp.data[field])}`);
      }
    }

    // age_ms_present
    if (expect.age_ms_present !== undefined) {
      const actual = resp.ageMs !== null;
      if (actual !== expect.age_ms_present) {
        failures.push(`age_ms_present=${expect.age_ms_present} expected but got ${actual}`);
      }
    }

    // stale
    if (expect.stale !== undefined) {
      if (resp.stale !== expect.stale) {
        failures.push(`stale=${expect.stale} expected but got ${resp.stale}`);
      }
    }

    // error_contains
    if (expect.error_contains !== undefined) {
      const actual = resp.error ?? '';
      if (!actual.includes(expect.error_contains)) {
        failures.push(`error_contains=${JSON.stringify(expect.error_contains)} expected but error was ${JSON.stringify(actual)}`);
      }
    }

    return failures;
  }

  // ---------------------------------------------------------------------------
  // Main runner
  // ---------------------------------------------------------------------------

  async function main() {
    const fixtures = discoverFixtures();
    if (fixtures.length === 0) {
      console.error(`ERROR: no fixtures found under ${CONFORMANCE_DIR}`);
      process.exit(1);
    }

    console.log(`Found ${fixtures.length} conformance fixture(s).`);
    console.log(`Using COMB_BIN: ${COMB_BIN}\n`);

    const failures = [];
    let passed = 0;

    for (const fixture of fixtures) {
      let daemon = null;
      try {
        daemon = await spawnDaemon();
        const client = new Client({ socketPath: daemon.sockPath, timeoutMs: 5000 });

        // Run setup ops, ignore their responses.
        for (const setupOp of fixture.setup) {
          await runOp(client, setupOp);
        }

        // Run the test op and check expectations.
        const resp = await runOp(client, fixture.test);
        const expectFailures = checkExpect(fixture, resp);

        if (expectFailures.length > 0) {
          failures.push({
            name: fixture.name,
            description: fixture.description,
            sourcePath: fixture.sourcePath,
            reasons: expectFailures,
          });
          console.log(`  FAIL [${fixture.name}]`);
          for (const reason of expectFailures) {
            console.log(`       ${reason}`);
          }
        } else {
          passed++;
          console.log(`  pass [${fixture.name}]`);
        }
      } catch (err) {
        failures.push({
          name: fixture.name,
          description: fixture.description,
          sourcePath: fixture.sourcePath,
          reasons: [`unexpected exception: ${err instanceof Error ? err.message : String(err)}`],
        });
        console.log(`  FAIL [${fixture.name}] unexpected exception: ${err instanceof Error ? err.message : String(err)}`);
      } finally {
        if (daemon) daemon.stop();
      }
    }

    console.log(`\nResults: ${passed}/${fixtures.length} passed.`);

    if (failures.length > 0) {
      console.error(`\n${failures.length} conformance failure(s):`);
      for (const f of failures) {
        console.error(`\n  [${f.name}] ${f.description}`);
        console.error(`  path: ${f.sourcePath}`);
        for (const reason of f.reasons) {
          console.error(`  - ${reason}`);
        }
      }
      process.exit(1);
    }

    console.log('All conformance fixtures passed.');
    process.exit(0);
  }

  main().catch((err) => {
    console.error(`FATAL: ${err instanceof Error ? err.stack : String(err)}`);
    process.exit(1);
  });
})();
